scene/animation/skeleton_ik.h

```diff
@@ -0,0 +1,202 @@
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef SKELETON_IK_H
+#define SKELETON_IK_H
+
+/**
+ * @author AndreaCatania
+ */
+
+#include "core/math/transform.h"
+#include "scene/3d/skeleton.h"
+
+class FabrikInverseKinematic {
+
+	struct EndEffector {
+		BoneId tip_bone;
+		Transform goal_transform;
+	};
+
+	struct ChainItem {
+
+		Vector<ChainItem> childs;
+		ChainItem *parent_item;
+
+		// Bone info
+		BoneId bone;
+		PhysicalBone *pb;
+
+		real_t length;
+		/// Positions relative to root bone
+		Transform initial_transform;
+		Vector3 current_pos;
+		// Direction from this bone to child
+		Vector3 current_ori;
+
+		ChainItem() :
+				parent_item(NULL),
+				bone(-1),
+				pb(NULL),
+				length(0) {}
+
+		ChainItem *find_child(const BoneId p_bone_id);
+		ChainItem *add_child(const BoneId p_bone_id);
+	};
+
+	struct ChainTip {
+		ChainItem *chain_item;
+		const EndEffector *end_effector;
+
+		ChainTip() :
+				chain_item(NULL),
+				end_effector(NULL) {}
+
+		ChainTip(ChainItem *p_chain_item, const EndEffector *p_end_effector) :
+				chain_item(p_chain_item),
+				end_effector(p_end_effector) {}
+
+		ChainTip(const ChainTip &p_other_ct) :
+				chain_item(p_other_ct.chain_item),
+				end_effector(p_other_ct.end_effector) {}
+	};
+
+	struct Chain {
+		ChainItem chain_root;
+		Vector<ChainTip> tips;
+	};
+
+public:
+	struct Task : public RID_Data {
+		RID self;
+		Skeleton *skeleton;
+
+		Chain chain;
+
+		// Settings
+		real_t min_distance;
+		int max_iterations;
+
+		// Bone data
+		BoneId root_bone;
+		Vector<EndEffector> end_effectors;
+
+		Transform goal_global_transform;
+
+		Task() :
+				skeleton(NULL),
+				min_distance(0.01),
+				max_iterations(10),
+				root_bone(-1) {}
+	};
+
+private:
+	/// Init a chain that starts from the root to tip
+	static void build_chain(Task *p_task, bool p_force_simple_chain = true);
+
+	static void update_chain(const Skeleton *p_sk, ChainItem *p_chain_item);
+
+	/// Special solvers that solve only chains with one end effector
+	static void solve_simple_backwards(Chain &r_chain);
+	static void solve_simple_forwards(Chain &r_chain);
+
+public:
+	static Task *create_simple_task(Skeleton *p_sk, BoneId root_bone, BoneId tip_bone, const Transform &goal_transform);
+	static void free_task(Task *p_task);
+	// The goal of chain should be always in local space
+	static void set_goal(Task *p_task, const Transform &p_goal);
+	static void make_goal(Task *p_task, const Transform &p_inverse_transf, real_t blending_delta);
+	static void solve(Task *p_task, real_t blending_delta);
+};
+
+class SkeletonIK : public Node {
+	GDCLASS(SkeletonIK, Node);
+
+	String root_bone;
```

Comment:
by the way, I think this should be StringName, not String

Reply:
I changed it, but seems to me that it is not better, because the skeleton internally store bone names as String

Reply:
I've tried to set StringName anyway, but I got compilation error..